### ISSUE TITLE
Support multiple reporters in browser

### ISSUE THREAD

**Issue by letsgotomars** [mochajs/mocha#3902]
**Is your feature request related to a problem or a nice-to-have?? Please describe.**
A clear and concise description of what the problem is. E.g. I'm always frustrated when [...]

I think it's awesome that mocha supports running in the browser!

I would like mocha be able to support multiple reporters for the browser - mainly for using 'json-stream' with 'html'. Currently you can only specify one reporter.

**Describe the solution you'd like**

My first thought would be to have mocha.setup() be able to accept an array reporters instead of just a single string
For example:
`
    mocha.setup({
        ui: 'bdd',
        reporter: ['html', 'json-stream']
    });
`
Another solution would be to have json-stream as an additional option
`
    mocha.setup({
        ui: 'bdd',
        reporter: 'html',
        jsonStream: true,
    });
`
As a side note I already have a one-line hack working so I know it's definitely possible

**Describe alternatives you've considered**

**Additional context**

**Comment by janiukjf** [comment]
I would love the former option, especially for dealing with code coverage, nyc, and artifact output. I want my CI runs to show the results as they are running but also output to an xml file. I've tried other hacky npm packages, like mocha-multi, with mixed results and it would be nice if this was first party.

**Comment by letsgotomars** [comment]
@janiukjf code coverage is exactly why we did this modification. We have a tool that runs front-end mocha tests on istanbul instrumented code and then displays the test results using the Mocha 'html' reporter (with some custom dom manipulation). However the tool also hooks into the console logs from 'json-stream' to detect when tests are finished. When it detects the tests are finished it then generates [nyc html](https://istanbul.js.org/docs/advanced/alternative-reporters/#html) coverage files and appends a link to those files in the header of the Mocha 'html' report (the user can also optionally download an lcov file).

This allows users to run front-end mocha tests and immediately be able to see the test results and coverage all in one place. Attached is a screenshot of the tool after a test run. (The tool can also run multiple Mocha tests in parallel with remote controlled [Selenium](https://en.wikipedia.org/wiki/Selenium_(software)) headless-browsers to allow for front-end testing to be integrated into our CI pipline.) Thought I would share this use-case to demonstrate that there a lot of utility to be potentially gained with this feature.

![image](https://user-images.githubusercontent.com/26883069/59408368-85641a00-8d79-11e9-85ea-aeb46503d54b.png)

**Comment by antomor** [comment]
Any news on it? @letsgotomars could you share the one line hack necessary to achieve such result?

**Comment by letsgotomars** [comment]
Sure - I modified a large bundled mocha.js script that I got from a CDN. Sorry I can't remember the specific version we used in our tool. Here is the line I modified:

```js
    Mocha.prototype.run = function(fn) {
      if (this.files.length) {
        this.loadFiles();
      }
      var suite = this.suite;
      var options = this.options;
      options.files = this.files;
      var runner = new exports.Runner(suite, options.delay);
      createStatsCollector(runner);
      var reporter = new this._reporter(runner, options);
      // ===== < CUSTOM MOD > =====
      // Run 'json-stream' reporter always
      if (this._reporter != builtinReporters['JSONStream']) builtinReporters['JSONStream'](runner, options);
      // ===== </ CUSTOM MOD > =====
      runner.ignoreLeaks = options.ignoreLeaks !== false;
      runner.fullStackTrace = options.fullStackTrace;
      runner.asyncOnly = options.asyncOnly;
      runner.allowUncaught = options.allowUncaught;
      runner.forbidOnly = options.forbidOnly;
      runner.forbidPending = options.forbidPending;
      if (options.grep) {
        runner.grep(options.grep, options.invert);
      }
      if (options.globals) {
        runner.globals(options.globals);
      }
      if (options.growl) {
        this._growl(runner);
      }
      if (options.useColors !== undefined) {
        exports.reporters.Base.useColors = options.useColors;
      }
      exports.reporters.Base.inlineDiffs = options.useInlineDiffs;
      exports.reporters.Base.hideDiff = options.hideDiff;
    
      function done(failures) {
        fn = fn || utils.noop;
        if (reporter.done) {
          reporter.done(failures, fn);
        } else {
          fn(failures);
        }
      }
    
      return runner.run(done);
    };
```

**Comment by letsgotomars** [comment]
Also I made another one-liner change here:

```js
    Progress.prototype.update = function(n) {
      // ===== < CUSTOM MOD > =====
      // Provide hook into mocha testing progress
      window.mocha.percentComplete = n;
      // ===== </ CUSTOM MOD > =====
      this.percent = n;
      return this;
    };
```

**Comment by mostafaNazari702** [comment]
Coming to this one cuz i was tasked to get this done, and i see that there’s already like six closed PRs going way back (2015 and up, #930, #1360, #1772, #2091, #2184, and that partial #3234), althought not directly linked here all of them, it requires cross-referencing and of course, ✨reading✨. Before i open yet another one I would rather just sync with you guys first so we don't keep looping on this.

From reading through those old threads, it kinda feels like the main issue was not "do we need this" but more like nobody could really agree on what the shape should look like.

Every PR went a slightly different way. some tried comma-separated strings, some used arrays, some tried stuffing stuff into the reporter string itself like --reporter html:out.html. It all sorta works in isolation but it ends up pulling the CLI in different directions and nothing sticks.

Another thing that kept coming up is per reporter config. Once you got more than one reporter, you kinda _need_ a way to give each one its own options. But the old --reporter-options is global so stuff like xunit/json/markdown all end up fighting over the same config space. I think that's probly where a lot of the conflict came from.

Also the whole "done(failures, fn)" thing. Some reporters do async stuff (write files, touch DOM in browser, etc....), so they need time to finish. But when you got multiple reporters doing that it's not super clear who mocha is supposed to wait on or in what order. None of the older PRs really nailed that down clean ( i could be wrong, and i am open to be corrected but this is my interpretation for now ).

And then there's browser vs node parity. Even tho this issue is kinda browser focused, people still expect the same config shape in node, so whatever solution we pick kinda has to work in both places. The browser "mocha.setup({ reporter: 'html' })" and the node "Mocha.prototype.reporter" path should end up meaning the same thing not two diff models.

What I am thinking is something like this: instead of just one reporter, let it be an array. So you can pass multiple reporters, each with a name (or ctor) and optional options. Like html + json-stream at the same time, each doing their own thing. Backwards compat stays, so string/ctor still works like today but internally it just normalizes into a list.

BTS, mocha would just create each reporter and attach them all to the same runner. Since runner already emits events, multiple listeners already works fine so no big rewiring needed there.

On CLI side, you could repeat --reporter to build up the list, and then have something like namespaced options so each reporter gets its own config. Old single reporter usage stays exactly the same so nobody gets broken.

The tricky bit is still "done". my idea would be mocha just waits for all reporter "done" methods if they exist, kinda in parallel, and ignores ones that don't have it. Once all of them finish, mocha continues like normal. exit code still just comes from failures like before so behavior doesnt randomly change just cuz you added more reporters.

now migration wise, single reporter users shouldn't see any diff at all. even internal stuff that reads the reporter still gets the first one so nothing weird breaks for existing code that shouldn't be touching it anyway.
I now want to just check alignment on 3 things:

1. array of reporters instead of string/CSV style.
2. namespaced --reporter-option name.key=value style
3 waiting on multiple reporter done() calls like described

If those are ok i can cook a PR, if not it's better to know now than end up with another closed attempt.
